File: v8total/openvino/main.cpp
#include <iostream>
#include <fstream>
#include <opencv2/imgproc.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/dnn.hpp>
#include <openvino/openvino.hpp>


class YOLOV8OPENVINO
{
public:
	YOLOV8OPENVINO(std::string modelxml, float confThreshold, float nmsThreshold);
	void drawPred(cv::Mat& frame, cv::Rect box, float conf, int classid);
	cv::Mat resize_img(cv::Mat& image, int *neww, int*newh, int *padw, int *padh);
	void detect(cv::Mat& frame);
private:
	int inpWidth;
	int inpHeight;

	float confThreshold;
	float nmsThreshold;
	
	ov::InferRequest infer_request;
	ov::CompiledModel compiled_model_detect;
	
	int numproposal;
	int nout;

	std::vector<std::string> class_names;

};


YOLOV8OPENVINO::YOLOV8OPENVINO(std::string modelxml, float confThreshold, float nmsThreshold)
{
	this->confThreshold = confThreshold;
	this->nmsThreshold = nmsThreshold;
	
	this->inpWidth = 640;
	this->inpHeight = 640;

	std::ifstream ifs("coco.names");
	std::string line;

	while(getline(ifs, line)) this->class_names.push_back(line.substr(0, line.length()-1));
	
	this->nout = 84;
	this->numproposal = 8400;

	ov::Core core;

	this->compiled_model_detect = core.compile_model(modelxml, "CPU");

	this->infer_request = this->compiled_model_detect.create_infer_request();

}


void YOLOV8OPENVINO::drawPred(cv::Mat& frame, cv::Rect box, float conf, int classid)
{
	cv::rectangle(frame, box, cv::Scalar(123,45,67), 2);

	std::string label = cv::format("%.2f", conf);

	label = this->class_names[classid] + ":" + label;

	cv::putText(frame, label, cv::Point(box.x, box.y-10), cv::FONT_HERSHEY_SIMPLEX, 1, cv::Scalar(90,123,89), 2);

}

cv::Mat YOLOV8OPENVINO::resize_img(cv::Mat& image, int *neww, int *newh, int *padw, int *padh)
{
	*neww = this->inpWidth, *newh = this->inpHeight;

	int srcw = image.cols, srch = image.rows;

	cv::Mat simg;

	if(srcw != srch)
	{
		float hw_scale = (float)srch / srcw;
		if(hw_scale > 1.0)
		{
			*neww = int(this->inpWidth / hw_scale);
			cv::resize(image, simg, cv::Size(*neww, *newh), cv::INTER_AREA);
			*padw = int((this->inpWidth - *neww) * 0.5);
			cv::copyMakeBorder(simg, simg, 0, 0, *padw, this->inpWidth-*neww-*padw, cv::BORDER_CONSTANT, 114);
		
		}
		else
		{
			*newh = int(this->inpHeight * hw_scale);
			cv::resize(image, simg, cv::Size(*neww, *newh), cv::INTER_AREA);
			*padh = int((this->inpHeight - *newh) * 0.5);
			cv::copyMakeBorder(simg, simg, *padh, this->inpHeight-*newh-*padh, 0, 0, cv::BORDER_CONSTANT, 114);
		}
	
	
	}
	else
	{
		cv::resize(image, simg, cv::Size(*neww, *newh), cv::INTER_AREA);
	
	}

	return simg;

}

void YOLOV8OPENVINO::detect(cv::Mat& frame)
{
	int neww =0, newh = 0, padw = 0, padh = 0;
	cv::Mat simg = this->resize_img(frame, &neww, &newh, &padw, &padh);
	cv::Mat blob = cv::dnn::blobFromImage(simg, 1.0 / 255.0, cv::Size(this->inpHeight, this->inpWidth), cv::Scalar(0,0,0), true, false);

	auto input_port = this->compiled_model_detect.input();
	
	ov::Tensor inputTensor(input_port.get_element_type(), input_port.get_shape(), blob.ptr(0));	

	
	this->infer_request.set_input_tensor(inputTensor);
	

	this->infer_request.infer();

	auto output = infer_request.get_output_tensor(0);
	float *outdata = output.data<float>();
	cv::Mat output_buffer(this->nout, this->numproposal, CV_32F, outdata);
	cv::transpose(output_buffer, output_buffer);
	float* pdata = (float*)output_buffer.data;

	
	std::vector<cv::Rect> boxes;
	std::vector<int> ids;
	std::vector<float> confs;

	float ratioh = (float)frame.rows / newh;
	float ratiow = (float)frame.cols / neww;

	for(int n = 0; n < this->numproposal; ++n)
	{
		float maxss = 0.0;
		int idp = 0;
		for(int k = 0; k < this->nout - 4; ++k)
		{
			if(pdata[k + 4] > maxss)
			{
				maxss = pdata[k + 4];
				idp = k;
			}
		
		}

		if(maxss >= this->confThreshold)
		{
			float cx = (pdata[0] - padw) * ratiow;
			float cy = (pdata[1] - padh) * ratioh;
			float w = pdata[2] * ratiow;
			float h = pdata[3] * ratioh;

			int left = int(cx - 0.5 * w);
			int top = int(cy - 0.5 * h);

			boxes.push_back(cv::Rect(left, top, int(w), int(h)));
			confs.push_back(maxss);
			ids.push_back(idp);
		
		}
		pdata += this->nout;
	
	}
	
	std::vector<int> indices;
	cv::dnn::NMSBoxes(boxes, confs, this->confThreshold, this->nmsThreshold, indices);

	for(int idx : indices) this->drawPred(frame,  boxes[idx], confs[idx], ids[idx]);

}


int main()
{
	YOLOV8OPENVINO net("yolov8n_openvino_model/yolov8n.xml", 0.7, 0.8);

	cv::Mat srcimg = cv::imread("imgs/bus.jpg");

	net.detect(srcimg);

	cv::imwrite("imgs/bus_openvino_cpp.jpg", srcimg);

	return 0;
}

File: v8total/openvino/main.py
'''
pip install openvino

yolo export model=yolov8n.pt format=openvino
'''
import cv2
import numpy as np

from openvino.runtime import Core


class OpenVYolov8:

    def __init__(self, conf_thres=0.7, nms_thres=0.8):
        
        self.confThreshold = conf_thres
        self.nmsThreshold = nms_thres

        self.model_xml = "yolov8n_openvino_model/yolov8n.xml"
        self.model_bin = "yolov8n_openvino_model/yolov8n.bin"

        self.class_names = [x.strip() for x in open("coco.names", "r").readlines()]

        self.ie = Core()
        self.net = self.ie.read_model(model=self.model_xml, weights=self.model_bin)
        self.device = "CPU"
        self.complied_model = self.ie.compile_model(model=self.net, device_name=self.device)
        self.output_layer = self.complied_model.output(0)

        self.inpWidth = 640
        self.inpHeight = 640


    def drawPred(self, image, box, conf, clsid):

        x,y,w,h = box

        cv2.rectangle(image, (x,y), (x+w,y+h), (178,45,233), 2)

        label = self.class_names[clsid] + ":" + str(round(conf * 100)) + "%"

        cv2.putText(image, label, (x,y-10), cv2.FONT_HERSHEY_SIMPLEX, 1, (233,0,45), 2)

    def resize_img(self, image):

        neww, newh, padw, padh = self.inpWidth, self.inpHeight, 0, 0
        srch, srcw = image.shape[:2]
        if srch != srcw:
            hw_scale = srch / srcw
            if hw_scale > 1.0:
                neww = int(self.inpWidth / hw_scale)
                simg = cv2.resize(image, (neww, newh), cv2.INTER_AREA)
                padw = int((self.inpWidth - neww) * 0.5)
                simg = cv2.copyMakeBorder(simg, 0, 0, padw, self.inpWidth-neww-padw, cv2.BORDER_CONSTANT, (114,114,114))
            else:
                newh = int(self.inpHeight * hw_scale)
                simg = cv2.resize(image, (neww, newh), cv2.INTER_AREA)
                padh = int((self.inpHeight - newh) * 0.5)
                simg = cv2.copyMakeBorder(simg, padh, self.inpHeight-newh-padh, 0, 0, cv2.BORDER_CONSTANT, (114,114,114))
        else:
            simg = cv2.resize(image, (neww, newh), cv2.INTER_AREA)

        return simg, neww, newh, padw, padh


    def detect(self, image):

        simg, neww, newh, padw, padh = self.resize_img(image)
        blob = cv2.dnn.blobFromImage(simg, 1.0 / 255.0, swapRB=True)

        results = self.complied_model([blob])[self.output_layer]

        ratiow, ratioh = image.shape[1] / neww, image.shape[0] / newh
        boxes, confs, ids = [], [], []

        for pred in np.squeeze(results).transpose():

            maxss = np.max(pred[4:])
            idp = np.argmax(pred[4:])

            if maxss >= self.confThreshold:
                cx = (pred[0] - padw) * ratiow
                cy = (pred[1] - padh) * ratioh
                w = pred[2] * ratiow
                h = pred[3] * ratioh

                left = int(cx - 0.5 * w)
                top = int(cy - 0.5 * h)

                boxes.append([left, top, int(w), int(h)])
                confs.append(maxss)
                ids.append(idp)

        indices = cv2.dnn.NMSBoxes(boxes, confs, self.confThreshold, self.nmsThreshold)

        for idx in indices:
            self.drawPred(image, boxes[idx], confs[idx], ids[idx])
        return image

if __name__ == "__main__":
    
    srcimg = cv2.imread("imgs/bus.jpg")

    yolonet = OpenVYolov8()

    tarimg = yolonet.detect(srcimg)

    cv2.imwrite("imgs/bus_openvino.jpg", tarimg)


        
